File: crates/certimining-log/src/promise.rs
//! Signed inclusion promises, and what they can be used to show (§1.6, §2.3).
//!
//! A promise is the anti-censorship device: it binds one leaf to one epoch window under the batcher's
//! key, and a counterparty checks it without the batcher. What it cannot do is prove absence. Only the
//! batcher can show a leaf is present, so an unsatisfied promise is a transferable accusation the
//! batcher may rebut with an inclusion proof, and a rebuttal counts only if that proof resolves to a
//! root published inside the window. A proof against a later root confirms the breach (D-72).
//!
//! One difference from a record's attestation is worth stating, because getting it wrong would be
//! invisible: the qualified person signs the leaf **preimage**, while §1.6 signs the SPI **digest**.

use certimining_core::{
    Digest, Hasher, Preimage, RegistryError, Result, SpiPreimage, SubmissionId, Verifier,
};

use crate::{InclusionProof, InclusionVerifier, ProofVerifier};

/// §1.8's merge delay: the promised epoch and the two after it.
pub const MAX_MERGE_DELAY: u8 = 2;

/// A batcher's promise (§1.6, D-68).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SignedPromise {
    /// The exact leaf the promise binds (INV-SPI-02).
    pub leaf: Digest,
    /// The submission's ordinal identifier, which travels only inside this promise (D-69).
    pub submission_id: SubmissionId,
    /// The first epoch whose root may carry the leaf.
    pub promised_epoch: u64,
    /// How many epochs past `promised_epoch` still satisfy it.
    pub max_merge_delay: u8,
    /// The public half of the key that signed this promise, carried for display. Never the authority.
    pub batcher_key: [u8; 32],
    /// Ed25519 over the SPI digest, which is what §1.6 signs.
    pub signature: [u8; 64],
}

/// The digest §1.6 signs:
/// `Keccak256( TAG_SPI ‖ leaf ‖ submission_id ‖ promised_epoch ‖ max_merge_delay )`.
pub fn promise_digest<H: Hasher>(promise: &SignedPromise) -> Result<Digest> {
    SpiPreimage {
        leaf: promise.leaf,
        submission_id: promise.submission_id,
        promised_epoch: promise.promised_epoch,
        max_merge_delay: promise.max_merge_delay,
    }
    .digest::<H>()
}

/// Checks a promise against the batcher key the counterparty expects (D-68).
///
/// The key a promise carries is not the authority: anyone can sign a promise, so the expected key
/// decides first. A key that is not the expected one is `0x08`, decided before any verification runs,
/// exactly as §1.3's condition (c) decides a qualified person's key. A signature that does not verify
/// is `0x07`.
pub fn verify_promise<H: Hasher, V: Verifier>(
    promise: &SignedPromise,
    expected_batcher_key: &[u8; 32],
) -> Result<()> {
    if &promise.batcher_key != expected_batcher_key {
        return Err(RegistryError::AttestationKeyMismatch);
    }
    let digest = promise_digest::<H>(promise)?;
    V::verify(expected_batcher_key, &digest, &promise.signature)
}

/// Whether a promise was kept.
///
/// The proof must verify for the promise's own leaf against `root`, and `root_epoch` must fall inside
/// the promised window. A root published later than the window confirms the breach rather than
/// rebutting it, and returns `0x14`; a root earlier than the promise is outside the window too. A proof
/// that does not verify is `0x13`, which is also what a proof of some other leaf gives.
pub fn promise_kept<H: Hasher>(
    promise: &SignedPromise,
    proof: &InclusionProof,
    root: &Digest,
    root_epoch: u64,
) -> Result<()> {
    let last = promise
        .promised_epoch
        .checked_add(u64::from(promise.max_merge_delay))
        .ok_or(RegistryError::ArithmeticOverflow)?;
    if root_epoch < promise.promised_epoch || root_epoch > last {
        return Err(RegistryError::MergeDelayExceeded);
    }
    ProofVerifier::verify::<H>(&promise.leaf, proof, root)
}

File: ts/src/promise.ts
/**
 * §1.6's signed inclusion promise (SPI): its digest, its wire form, the policy §1.6 fixes, and
 * D-72's rule about which roots can rebut an accusation.
 */
import { type Digest, bytesEqual, concat, readU64le, u64le } from "./bytes.ts";
import { RegistryFailure } from "./errors.ts";
import { ed25519Verify, keccak256 } from "./hash.ts";
import { spiPreimage } from "./preimage.ts";
import { type InclusionProof, verifyInclusion } from "./tree.ts";

/** INV-SPI-01 fixes it at 2 epochs. The key holder does not choose the policy it is judged against. */
export const MAX_MERGE_DELAY = 2;

export type SignedPromise = {
  leaf: Digest;
  submissionId: Uint8Array;
  acceptedEpoch: bigint;
  promisedEpoch: bigint;
  maxMergeDelay: number;
  /** Carried for display, never the authority: `verifyPromise` takes the key the counterparty expects. */
  batcherKey: Uint8Array;
  signature: Uint8Array;
};

export type PublishedRoot = {
  epoch: bigint;
  root: Digest;
};

/** The 73-byte preimage of §1.6, and the digest the batcher actually signs. */
export function spiDigest(p: SignedPromise): Digest {
  return keccak256(
    spiPreimage({
      leaf: p.leaf,
      submissionId: p.submissionId,
      acceptedEpoch: p.acceptedEpoch,
      promisedEpoch: p.promisedEpoch,
      maxMergeDelay: p.maxMergeDelay,
    }),
  );
}

export const PROMISE_ENCODED_LEN = 161;

/** Borsh over §2.3's `SignedPromise` in its declared field order. */
export function encodePromise(p: SignedPromise): Uint8Array {
  return concat([
    p.leaf,
    p.submissionId,
    u64le(p.acceptedEpoch),
    u64le(p.promisedEpoch),
    Uint8Array.of(p.maxMergeDelay),
    p.batcherKey,
    p.signature,
  ]);
}

export function decodePromise(b: Uint8Array): SignedPromise {
  if (b.length !== PROMISE_ENCODED_LEN) {
    throw new RegistryFailure("MalformedPayload", `a promise is ${PROMISE_ENCODED_LEN} bytes, got ${b.length}`);
  }
  return {
    leaf: b.slice(0, 32),
    submissionId: b.slice(32, 48),
    acceptedEpoch: readU64le(b, 48),
    promisedEpoch: readU64le(b, 56),
    maxMergeDelay: b[64]!,
    batcherKey: b.slice(65, 97),
    signature: b.slice(97, 161),
  };
}

/**
 * §2.3 `verify_promise`. A key mismatch is 0x08, decided first. A signature proves authorship and
 * not compliance, so a `max_merge_delay` other than INV-SPI-01's, a `promised_epoch` outside the
 * window the signed `accepted_epoch` allows, or an `accepted_epoch` that is not the observed epoch
 * or exactly one behind it, is 0x17. A signature that does not verify is 0x07.
 *
 * `observedEpoch` is the caller's own reading of the checkpoint sequence when the promise arrived.
 * Nothing in the artifact can supply it (§1.6's three statements), so it is an input.
 */
export function verifyPromise(promise: SignedPromise, expectedBatcherKey: Uint8Array, observedEpoch: bigint): void {
  if (!bytesEqual(promise.batcherKey, expectedBatcherKey)) {
    throw new RegistryFailure("AttestationKeyMismatch", "the promise names a batcher key the counterparty does not expect");
  }
  if (promise.maxMergeDelay !== MAX_MERGE_DELAY) {
    throw new RegistryFailure("PromisePolicyInvalid", `max_merge_delay ${promise.maxMergeDelay}, and INV-SPI-01 fixes it at ${MAX_MERGE_DELAY}`);
  }
  // Backdating: an acceptance epoch later than observed lets a batcher defer its own
  // accountability. One epoch behind is allowed because a promise can cross an epoch boundary.
  if (promise.acceptedEpoch > observedEpoch || promise.acceptedEpoch + 1n < observedEpoch) {
    throw new RegistryFailure("PromisePolicyInvalid", `accepted_epoch ${promise.acceptedEpoch} against an observed ${observedEpoch}`);
  }
  if (
    promise.promisedEpoch < promise.acceptedEpoch ||
    promise.promisedEpoch > promise.acceptedEpoch + BigInt(promise.maxMergeDelay)
  ) {
    throw new RegistryFailure("PromisePolicyInvalid", `promised_epoch ${promise.promisedEpoch} is outside the accepted window`);
  }
  if (!ed25519Verify(promise.batcherKey, spiDigest(promise), promise.signature)) {
    throw new RegistryFailure("AttestationInvalid", "the promise's signature does not verify over the SPI digest");
  }
}

/**
 * §2.3 `promise_kept`. The proof must be for the same epoch as the root it is checked against,
 * that epoch must fall inside the promised window, and the path must verify for the promise's own
 * leaf. Outside the window in either direction is 0x14 (D-72); an inconsistent or failing proof is
 * 0x13. It does not establish that the root was published: provenance is the caller's seam.
 */
export function promiseKept(promise: SignedPromise, proof: InclusionProof, published: PublishedRoot): void {
  if (proof.epoch !== published.epoch) {
    throw new RegistryFailure("InclusionProofInvalid", "the proof is for a different epoch than the root");
  }
  const windowEnd = promise.promisedEpoch + BigInt(promise.maxMergeDelay);
  if (published.epoch < promise.promisedEpoch || published.epoch > windowEnd) {
    throw new RegistryFailure("MergeDelayExceeded", `epoch ${published.epoch} is outside ${promise.promisedEpoch}..${windowEnd}`);
  }
  verifyInclusion(promise.leaf, proof, published.root);
}
